### ISSUE TITLE
Cascade Delete for Relational Items (Foreign Key Constraints)

### ISSUE THREAD

**Issue by benhaynes** [directus/directus#2942]
To avoid dead/stranded items, when you delete an item we should do a check if it is referenced elsewhere relationally, and if so, we should either NULLify or delete the relationship. Ideally all of this is within an API parameter that can be set on a case-by-case basis.

* **Example 1**: You delete a file, and therefore should NULLify all M2O references to that file (Ref: Ref: https://github.com/directus/next/issues/529)
* **Example 2**: You delete an item, and therefore should delete its O2M children translations (Ref: https://github.com/directus/next/issues/593)

**Comment by Oreilles** [comment]
> What about warning that by deleting this item, references get modified and then you can choose between set to null, delete and cancel. I would also be helpful, to at least see some relations shown, maybe all in another modal?

This adds overhead and would be error prone, plus the database already does that validation for free.
The best way to handle that would be to simply add an option within the delete popup card with `cancel (default)`, `cascade`, `nullify`.

This means that each relation created within directus should also add a constraint on the corresponding fields. Knex does that very well with `column.references(field).inTable(table)`. Doing so would make directus relations much safer.

I went through the api code but I'm still not so sure whether it would best fit in `relations.post` or `FieldsService.createField` / `addColumnToTable` ?

**Comment by rijkvanzanten** [comment]
> I went through the api code but I'm still not so sure whether it would fit best in relations.post or FieldsService.createField / addColumnToTable ?

This is a very good question, I'm not sure yet myself.

On the one hand, it makes sense to handle the key constraints when a field is created (which would be `FieldsService.createField` / `addColumnToTable`), but we also want to make sure you can retroactively attach a relationship to an existing field, so that would mean in `relations.post`. 

All in all, I think it makes most sense to keep foreign key constraints coupled to `directus_relations`. When a relation is deleted, we should also delete the foreign key constraint

**Comment by Oreilles** [comment]
If we do that just after
```javascript
const primaryKey = await service.create(req.body);
try {
    const item = await service.readByKey(primaryKey, req.sanitizedQuery);
```
then we need to deduce that it is a `m2o` relationship from the object props that aren't falsy

```javascript
item.many_collection &&
item.many_field &&
item.many_primary &&
item.one_collection &&
item.one_primary
```
, or we can fetch the `one_collection` table fields and check the `special` prop of `one_primary`.


> When a relation is deleted, we should also delete the foreign key constraint

Can relations actually be deleted ? Field types cannot change currently, so if you want to drop a relation you have to drop the field. If that field is depended on by a `m2o` in another table the database won't allow it (free check). Just like with items, this should be handle with an option from the request (`cancel`, `cascade`, `unbind`).

**Comment by rijkvanzanten** [comment]
> then we need to deduce that it is a m2o relationship from the object props that aren't falsy

The foreign key will always be `many_field` -> `one_collection` + `one_primary`. A o2m can't exist without a m2o being present, so that should give us enough security 🤔 

> Field types cannot be changed currently

They can though! We don't allow it in the app because it's almost a guarantee for data corruption, but the API supports it.

> so if you want to drop a relation you have to drop the field

This isn't necessarily true, similar to the "field type can't be changed" from above, you can still `DELETE` to `/relations` and do it.

**Comment by Oreilles** [comment]
> They can though! We don't allow it in the app because it's almost a guarantee for data corruption, but the API supports it.

That's interesting ! I don't get the point about data corruption though. How is data corruption more of a problem within the app that within the api ?

> This isn't necessarily true, similar to the "field type can't be changed" from above, you can still DELETE to /relations and do it.

I see. In that case the relation controller is the only possible place to handle the constraints creation and deletion.

Talking about that make me think that directus is essentially trying to reimplement table relations when it allready is, and all relations being already available in the database native tables (pg_catalog et al...) with all sort of problems like field changing names and table changing schema already handled. Why not leverage on the very feature of a rdbms ?

**Comment by rijkvanzanten** [comment]
> I don't get the point about data corruption though. How is data corruption more of a problem within the app that within the api ?

It's not! The risk is the same. It's a difference of "target audience". The application offers all sorts of handholding and guiding to make sure things operate smoothly, the API is more "at your own risk", but therefore also more flexible 🙂 

> Can we start a branch / pull request to work on that ?

Totally! I'm very swamped on getting bugs under control next week, but you're more than welcome to get something started.

> Why not leverage on the very feature of a rdbms ?

Not every database supports it properly, not every database returns that info reliably, Directus won't forever be running just relational databases, there's relationship types that aren't really possible in just foreign key constraints (like the many-to-any, many-to-many w/ junctions), it would be kinda slow to have to search for all FKs just to know if any given field is a relational o2m/m2m setup, etc

**Comment by rijkvanzanten** [comment]
The support in SQL Server is slightly different than that from other SQL vendors, so we should look into using Triggers instead of on-update/on-delete triggers for that vendor (see https://github.com/directus/directus/pull/4041/files#diff-ddd0c62a0d65f7876e0ced20254076c6649e3872c0fd40e3aebac0715bb56dd9)

**Comment by josephdpurcell** [comment]
Sorry if I missed something, but wanted to recap what I found in case someone else lands here and could benefit. I think this is the most recent issue on the topic.

Here's what I found:

* Cascade delete on relationships like m2a doesn't delete the item only the junction. See this ticket and several others:
  * https://github.com/directus/v8-archive/issues/384
  * https://github.com/directus/v8-archive/issues/873
  * https://github.com/directus/directus/discussions/11245#discussioncomment-6516625)
* The naming of the collection on the deselect action of m2a is changing to clarify that its the relational table that gets the cascade delete, not the item itself https://github.com/directus/directus/issues/17867
* If you need a cascade delete on the related item, implement a Flow; I did not see an example Flow for this
* I think file relationships work similarly? If you delete an item in a collection and want to delete associated files you need a Flow.

**Comment by areejnaqvi** [comment]
> Sorry if I missed something, but wanted to recap what I found in case someone else lands here and could benefit. I think this is the most recent issue on the topic.
> 
> Here's what I found:
> 
> * Cascade delete on relationships like m2a doesn't delete the item only the junction. See this ticket and several others:
>   
>   * [Cascading Relational Delete v8-archive#384](https://github.com/directus/v8-archive/issues/384)
>   * ["Cascade" relationship deletions when deleting a parent item v8-archive#873](https://github.com/directus/v8-archive/issues/873)
>   * [Relationship triggers do not cascade delete when parent collection is deleted (m2a) #11245 (comment)](https://github.com/directus/directus/discussions/11245#discussioncomment-6516625))
> * The naming of the collection on the deselect action of m2a is changing to clarify that its the relational table that gets the cascade delete, not the item itself [m2a delete constraints: missleading naming #17867](https://github.com/directus/directus/issues/17867)
> * If you need a cascade delete on the related item, implement a Flow; I did not see an example Flow for this
> * I think file relationships work similarly? If you delete an item in a collection and want to delete associated files you need a Flow.

hi @josephdpurcell, i was wondering if you've found a way to implement a flow for cascade delete?

**Comment by josephdpurcell** [comment]
@areejnaqvi I did not. BUT, I did discover something that I think will lend itself nicely to writing a flow, which is this:

1 - Edit the field --> Relationship --> Relational Triggers and use "Nullify the item field" for both

<details>
<summary>Screenshot</summary>
<img width="610" alt="Screen Shot 2023-09-13 at 12 04 54" src="https://github.com/directus/directus/assets/778111/28b80e98-fe42-4340-97a6-803044b70787">
</details>

2 - Cascade delete could then be implemented on a cron by looking at the relational table for a missing ID

For example, if Foo has a m2a to Bar theres a `foo_bar` relational table with a field `foo_uuid` (or some ID field for foo collection). The `foo_uuid` field would be null if that Foo entity was deleted or if Foo entity deselected the Bar item. Thus, any Bar items for that record can be deleted.

So I imagine the flow would be something like "query the `foo_bar` table where `foo_uuid` is null, for each row delete all Bar items matching the item ID in that row". If I do end up writing the flow I'll report back.
